### PR TITLE
Update I18n modules to support loading automatically bundled CLDR data

### PR DESCRIPTION
**Pull request by agubler**
## **Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code has been formatted with [`prettier`](https://prettier.io/) as per the [readme code style guidelines](./../#code-style)
* [x] Unit or Functional tests are included in the PR

### Description:

Comprehensive change of i18n support in Dojo; The motivation is to enable automatic detection of CLDR data requirements with lazy loading and resolution of i18n CLDR data bundles. Ensuring that CLDR data automatically included in i18n enables Dojo to remove existing custom locale resolution logic and entirely depend on `Globalize` for locale resolution.

This _should_ be backwards compatible, unless the downstream application is importing interfaces that are no longer needed (and can be removed) or using the core i18n module directly (which shouldn't be required).

### Details:

1) Expose all i18n message bundle and locale behaviour using reactive patterns from the existing i18n mixin and middleware. 
1) Expose an API for registering CLDR loaders, default locale and other supported locales.
1) Migrate to using `Globalize` for all message bundle resolution.
1) Lazily load required CLDR data using loaders registered with the i18n system.
1) Fallback to using a complete CLDR supplemental bundle when an unknown locale is used.
1) Basic i18n support for codesandbox using a util `cldr` module that loads the required supplemental data (This import will automatically be elided during the Dojo build)
1) Lazily loading CLDR data when updating a widget's locale through either the global application injector or explicitly via the locale property (support in both the middleware and mixin)

Related to #655

### CODE REVIEW

File: tests/i18n/support/cldr.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	loadData(enCurrencies);
+	loadData(enNumbers);
+	loadData(enUnits);
+	loadData(frCaGregorian);
+	loadData(frDateFields);
+	loadData(frTimeZoneNames);
+	loadData(frCurrencies);
+	loadData(frNumbers);
+	loadData(frUnits);
+});
```
Comment (mwistrand) [inline tests/i18n/support/cldr.ts]: Just verifying nothing throws?

File: tests/core/unit/middleware/i18n.tsx [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+			'<div><div>{"foo":"hello, {name}"}</div><div>hello, John</div><div>false</div><div>{}</div><button>es</button></div>'
+		);
+		root.children[0].children[4].click();
+		await localeLoader;
+		resolvers.resolveRAF();
+		assert.strictEqual(
+			root.innerHTML,
+			'<div><div>{"foo":""}</div><div></div><div>true</div><div>{"locale":"es"}</div><button>es</button></div>'
+		);
+		es.resolver({ default: { foo: 'holla, {name}' } });
```
[the remark this answers is not in the document]
Comment (agubler) [reply tests/core/unit/middleware/i18n.tsx]: Lol Thanks 😄

File: src/i18n/i18n.ts [hunk 3]
```diff
@@ -1,515 +1,367 @@
...
-
-	const cached = bundleMap.get(id);
-	const messages = cached ? cached.get(bundleLocale) || cached.get('root') : null;
-
-	if (!messages) {
-		throw new Error(`The bundle has not been registered.`);
+	const localCldrLoader = cldrLoaders[userLocale];
+	if (localCldrLoader && localCldrLoader !== true) {
+		loaderPromises.push(localCldrLoader());
+	}
```
[the remark this answers is not in the document]
Comment (agubler) [reply src/i18n/i18n.ts]: Yup absolutely

File: tests/i18n/support/cldr.ts (hunk 1 above)
Comment (agubler) [reply tests/i18n/support/cldr.ts]: This is used to load data in for the tests, runs in both browser and node.

File: src/i18n/i18n.ts [hunk 4]
```diff
@@ -1,515 +1,367 @@
...
+
+	return {
+		messages: Object.keys(bundle.messages).reduce(
+			(messages, key) => {
+				const message = globalize.cldr.get(`${MESSAGE_BUNDLE_PATH}/${bundleId}/${key}`);
+				messages[key] = message;
+				return messages;
+			},
+			{} as any
+		),
```
[the remark this answers is not in the document]
Comment (agubler) [reply src/i18n/i18n.ts]: Done.

**Comment by agubler** [comment]
@mwistrand I think I've addressed your comments

**Pull request merged**